server/routes/snapshotListener.ts

```diff
@@ -0,0 +1,97 @@
+import { Request, Response, NextFunction } from 'express';
+import { USERS } from 'construct-ui/lib/esm/components/icon/generated/IconNames';
+import { Test } from 'mocha';
+import { MixpanelSnapshotEvents } from 'shared/analytics/types';
+import { DB } from '../database';
+import { success } from '../types';
+import SnapshotSpaceCache from '../util/snapshotSpaceCache';
+import { mixpanelTrack } from '../util/mixpanelUtil';
+import { NotificationCategories } from '../../shared/types';
+
+
+const snapshotListener = async (
+  models: DB, cache: SnapshotSpaceCache, req: Request, res: Response, next: NextFunction
+  ) => {
+
+  // Check cache to see if this is a space we care about
+  const validSnapshot = await cache.check(req.body.space);
+
+  if (!validSnapshot) {
+    mixpanelTrack({
+      event: MixpanelSnapshotEvents.SNAPSHOT_INVALID_SPACE,
+      isCustomDomain: false,
+      space: req.body.space
+    });
+    return success(res, "Snapshot POST Recieved, Space not present");
+  }
+
+  const chainsToNotify = await models.Chain.findAll({
+    attributes: ['id'],
+    where: {
+      snapshot: [req.body.space]
+    },
+   });
+
+  // TODO: Figure out how to log the amount of times the listener is pinged
+  // mixpanel track
+
+  await chainsToNotify.forEach(async (e) => {
```

Comment:
1. `forEach` returns undefined, the proper way to do this would be `await Promise.all(chainsToNotify.map(() => ... )))`
2. Running database operations (notification creating) in a loop is extremely inefficient, we should avoid it if possible + figure out how to batch the notifications, if we can.

Reply:
1. Should be an easy fix, I will do this probably tomorrow

2. The loop is there mostly to get the chainInstances out of the array that the DB returns them in by default, even though a high percentage of the time (right now all the time) there will only be one element in the array. 
The only case in which this array will have more than one element is if multiple chains have the same snapshot space in the DB, in which case the amount of operations would just be the amount of chains subscribed to that space. Right now I don't think we have any spaces that multiple chains are linked to, but Zak mentioned that it could theoretically happen in the future so I figured that I should account for that now rather than later

Reply:
this loop should construct the notification object, but not emit it. the emission should happen for all notification objects constructed _after_ the loop, all at once in a bulk query.